caffe2/operators/reduce_ops.cc

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (c) 2016-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "caffe2/operators/reduce_ops.h"
+
+namespace caffe2 {
+
+// For a Tensor X of n dimensions (dims[0], ..., dims[ndim-1]), given index
+// is converted to corresponding n-dimensional index, e.g. for X.shape = (2,
+// 3, 4) the linear index 12 maps to 3-dimensional index (1, 0, 0).
+vector<TIndex> ConvertFromInputIndex(TIndex index, vector<TIndex>& dims) {
```

Comment:
Hey folks, this is pretty terrible performance to call the entire conversion (with vector allocation even, dang) on each element.

I highly doubt it makes sense to implement operators if their performance is that bad. Please do benchmark your ops and try to write more optimal implementation.

In this case we could actually proxy to Eigen - it seems to have an implementation for multi-axis reduce (thanks @jamesr66a for pointing it out)

Reply:
@dzhulgakov thanks for pointing out the problem. Yeah, we didn't do appropriate performance testing. Hi @zem7, @jamesr66a is working on this issue.